### PR TITLE
service: cleanup mempool and peer update shutdown

### PR DESCRIPTION
**Pull request by tychoish**

### CODE REVIEW

**Review by creachadair — APPROVED** [review]

**Review by williambanfield — APPROVED** [review]

**Comment by tychoish** [comment]
to avoid hitting a series of deadlocks in the tests, I ended up pushing a second batch of work to this branch, and it could probably use a quick pass.

File: internal/p2p/peermanager.go [hunk 1]
```diff
@@ -902,18 +885,15 @@ func (m *PeerManager) processPeerEvent(pu PeerUpdate) {
...
-		select {
-		case <-sub.doneCh:
-			continue
-		default:
+		if ctx.Err() != nil {
+			return
 		}
 		select {
+		case <-ctx.Done():
+			continue
```
Comment (cmwaters) [inline internal/p2p/peermanager.go:895]: why not just return here?

**Review by cmwaters — APPROVED** [review]
Niccceeee 👍  - is there a changelog somewhere already documenting the change to using contexts for lifecycle management?

File: internal/p2p/peermanager.go (hunk 1 above)
Comment (tychoish) [reply internal/p2p/peermanager.go:895]: I was trying to mirror the legacy behavior, but you’re right.

**Comment by tychoish** [comment]
> Niccceeee +1 - is there a changelog somewhere already documenting the change to using contexts for lifecycle management?

I'm pretty sure that the surface area of this (modulo changes to libs/service.Service,) is all internal, and the leak detector confirms that.

**Pull request merged**